### ISSUE TITLE
What is the `recipe` package?

### ISSUE THREAD

**Issue by albertz** [rwth-i6/i6_core#11]
I see that the code is using `import recipe.i6_core...`. What is the `recipe` package? Why should we do it this way? I thought the intention was to directly have `i6_core` in path, and then for `import i6_core...`? Having it in an implicit `recipe` package restricts the way this must be set up. And `recipe` is a package name which can much more easily clash with other packages. `i6_core` as the root package name would be much safer.

**Comment by curufinwe** [comment]
This is a piece of sisyphus "magic" as sisyphus will modify the loading of these modules in here:
https://github.com/rwth-i6/sisyphus/blob/master/sisyphus/loader.py#L210

I do not remember why we did it like this though. Maybe @critias remembers.

**Comment by albertz** [comment]
Yea but we could change this. I'm just afraid that we might run into problems at some point (e.g. whenever there is another `recipe` package, which is a very generic name).

Also when you want to use this outside of Sisyphus, this can look strange (it must be setup in this specific way to work).

**Comment by curufinwe** [comment]
Well, the recipies are somewhat bound to sisyphus anyway. How do you want to use them outside of it?

**Comment by albertz** [comment]
Even if only for Sisyphus, we might get these problems.

But also, maybe there are some parts which could be useful outside. Some generic utilities or whatever. Or maybe in some RETURNN config you directly import parts of the recipes.

**Comment by curufinwe** [comment]
If a piece of code is so useful it can be used outside these recipies, I'd argue it should be part of some other repo and imported into the recipies. In any case this issue is better addressed in the sisyphus discussions.

**Comment by JackTemaki** [comment]
I agree with Eugen, I think we discussed that the i6_core "recipe" repository should only contain (stable) Jobs and helpers that are directly related to those, so nothing should be needed outside Sisyphus. Also all other recipe packages follow exactly this scheme, so it would not be a good idea to change it here.

> Or maybe in some RETURNN config you directly import parts of the recipes.

Especially this should never be the case, because then we end up with similar utilities across multiple repos. The i6_core is for pipeline/workflow components, not for RETURNN helpers.

**Comment by albertz** [comment]
> > Or maybe in some RETURNN config you directly import parts of the recipes.
> 
> Especially this should never be the case, because then we end up with similar utilities across multiple repos. The i6_core is for pipeline/workflow components, not for RETURNN helpers.

There can be things which are useful for both. The point is exactly to not replicate similar utilities.

E.g. when I look at stuff like [this](https://github.com/rwth-i6/i6_core/blob/main/lib/lexicon.py), or other files there in `lib`, this looks pretty much not specific to Sisyphus jobs. We actually have a lot of this code already also as part of RETURNN. And then partly also as part of returnn-experiments/common. And I do use things like this in some of my RETURNN configs.

I don't see why we should restrict `i6_core` to only stuff for Sisyphus recipes. Sure, we can make separate repos for other things but not sure if such separation really makes it simpler. Ok, for public RETURNN specific things, we have returnn-experiments/common, but then what about things which are useful for both i6_core and returnn-experiments/common? Duplicate?

In any case, I just think this style of `import recipe...` is not so clear, no matter the other arguments. This imposes the implicit assumption that the package must be a subpackage in `recipe`.

**Comment by JackTemaki** [comment]
> I don't see why we should restrict i6_core to only stuff for Sisyphus recipes.

This is what we all agreed on, but of course we could open that discussion again.

> This imposes the implicit assumption that the package must be a subpackage in recipe.

This is a fixed Sisyphus convention, e.g.:

`global_constants.py`:
```
RECIPE_PREFIX = "recipe"
```

`job.py`:
```
        recipe_prefix = gs.RECIPE_PREFIX + '.'
        assert module_name.startswith(recipe_prefix)
```

So any objections about this should go into the Sisyphus repo.

**Comment by JackTemaki** [comment]
Maybe another clarification: RETURNN networks are not constructed/defined inside the i6_core. Because such helpers should be available for people not using Sisyphus. This would be part of the specific setup (which would be part of i6_private, i6_experiments or just in your local config), and there you would import returnn-experiments/common (which would also be possible via the new _import).

**Comment by albertz** [comment]
> > I don't see why we should restrict i6_core to only stuff for Sisyphus recipes.
> 
> This is what we all agreed on, but of course we could open that discussion again.

Did we? That was not my understanding. Or maybe I remember wrongly, or did not pay attention. I actually remember exactly the opposite, that we agreed that i6_core/i6_experiments would be the common place for all non-pure RETURNN stuff (which would go to returnn-experiments/common). The difference between i6_core vs i6_experiments is just that i6_core is meant to be stable, while i6_experiments is experimental, and more organized around users.

> > This imposes the implicit assumption that the package must be a subpackage in recipe.
> 
> This is a fixed Sisyphus convention, e.g.: ...
> So any objections about this should go into the Sisyphus repo.

Hm, maybe I don't know Sisyphus well enough, but why is it restricted? Why can't you also define `Job`s in other modules/packages which do not have this `recipe.` prefix?

>> We actually have a lot of this code already also as part of RETURNN
>
> Then we should discuss if we import RETURNN in the recipes (I am doing this for AppTek recipes), but not the other way around.

Well, yes, RETURNN itself should not depend on Sisyphus recipe stuff. But RETURNN configs or other external code (in recipes or tools or whatever) which make use of RETURNN could also use maybe things here in `i6_core`.

There are potentially also things which do not really make sense to be part of RETURNN but which are also not Sisyphus jobs. (I don't really have good examples currently, but I think there might be such things.)

**Comment by JackTemaki** [comment]
> Did we?

Thats how I understood it, and also how we wrote it into "CommonCode": "recipes/jobs. only stable, non experimental. hashes should not change, should not break old setups " and the slack channel: "non-breaking and tested asr recipes (mostly jobs, helpers only in extremely generic/well tested cases)"


> Hm, maybe I don't know Sisyphus well enough, but why is it restricted?

This is just how it was designed, "config" for pipeline definition, "recipe" for job definitions and "work" for the actual jobs. Having multiple packages was never intended, we just decided to do that now, but still follow the existing conventions.

>  But RETURNN configs or other external code (in recipes or tools or whatever) which make use of RETURNN could also use maybe things here in `i6_core`.

If this is really the case, you could still copy it to your config. But the `i6_core` is definitely not intended to be used as regular python package. Or at least the old "asr-recipes" repo from which the `i6_core` is derived from was never intended in that way. And the main goal was to have the asr-recipes clean and accessible to everyone.

**Comment by albertz** [comment]
> > Did we?
> 
> Thats how I understood it, and also how we wrote it into "CommonCode": "recipes/jobs. only stable, non experimental. hashes should not change, should not break old setups " and the slack channel: "non-breaking and tested asr recipes (mostly jobs, helpers only in extremely generic/well tested cases)"

I wrote that, and by recipes/helpers, I did not mean just Sisyphus recipes.

> > Hm, maybe I don't know Sisyphus well enough, but why is it restricted?
> 
> This is just how it was designed, "config" for pipeline definition, "recipe" for job definitions and "work" for the actual jobs. Having multiple packages was never intended, we just decided to do that now, but still follow the existing conventions.

We do not have to follow some restrictions which were made up earlier. We can always think about how we maybe can improve things. Esp when some conditions change.

I'm more asking if this would break anything. Or what would be needed to make it work. What would happen if we just remove that `assert` in Sisyphus?

**Comment by JackTemaki** [comment]
> I wrote that, and by recipes/helpers, I did not mean just Sisyphus recipes.

But I think this is was everybody else had in mind. This is really only about the `lib` folder, not about anything else, but lets discuss this next Tuesday.

> We do not have to follow some restrictions which were made up earlier

```
assert module_name.startswith(recipe_prefix)
sis_name = os.path.join(module_name[len(recipe_prefix):].replace('.', os.path.sep), cls.__name__)
```
=>
```
if module_name.startswith(recipe_prefix)
  sis_name = os.path.join(module_name[len(recipe_prefix):].replace('.', os.path.sep), cls.__name__)
else:
  sis_name = os.path.join(module_name.replace('.', os.path.sep), cls.__name__)
```
This is probably all that is needed, but of course this needs to be tested.

**Comment by christophmluscher** [comment]
Extending functionality of sisyphus resolved rwth-i6/sisyphus#52/pull/52

**Comment by albertz** [comment]
This is not really resolved. rwth-i6/sisyphus#52 now just allows for standard Python usage/code (which I expected anyway, this is why I opened this issue).

However, we still have a lot of `import recipe.i6_core...` in here (which is what this issue is about). This should be changed to `import i6_core...`.
